### PR TITLE
Create requests

### PR DESCRIPTION
**Pull request by mermop**
[Create a request with an existing user](https://trello.com/c/dAveFGU3/57-initiate-a-request-with-an-existing-user)
[Initiate a request by sending an email to someone new](https://trello.com/c/ITOS5m4u/58-initiate-a-request-by-sending-an-email-to-someone-who-doesnt-have-an-account)
[Requesters can see requests on their dashboard](https://trello.com/c/fyh8nlfH/63-requesters-can-see-requests-on-their-dashboard)
[Requestees can see requests on their dashboard](https://trello.com/c/YWMCeUwF/64-requestees-can-see-requests-on-their-dashboard)

## Tests that need to be written

Feature tests:
- [x] A request can be created
- [x] A request is shown on an existing user's dashboard
- [x] The number of unresolved requests is shown on a user's header

Controller tests:
- [x] Organisation member dashboard
- [x] Organisation member requests
- [x] User requests

Unit tests:
- [x] `Request.unresolved` include initiated and received requests but not cancelled, declined or responded ones
- [x] `user#find_or_invite` finds existing users and invites ones that don't exist

### CODE REVIEW

File: app/controllers/organisation_member/requests_controller.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  def new
+    @request = Request.new(requester: @organisation)
+  end
+
+  def create
+    @request = Request.new(document_type: request_params[:document_type],
+                           note: request_params[:note],
+                           requester: @organisation,
+                           requestee: User.find_or_invite(request_params[:requestee_email])
```
Comment (mermop) [inline app/controllers/organisation_member/requests_controller.rb]: There's a little bit of complexity here that I'm not dealing with:

- We probably shouldn't send an invitation to a user until we've saved the request. 
- But the request is invalid without a `User` as the requestee - so validation will fail if we move this later on.

I think the answer is probably conditional validation on the `Request` requirement for a requestee

File: app/models/request.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class Request < ApplicationRecord
+  belongs_to :requester, class_name: 'Organisation', foreign_key: 'requester_id'
+  belongs_to :requestee, class_name: 'User', foreign_key: 'requestee_id'
```
Comment (mermop) [inline app/models/request.rb]: Is requestee a bad name? `request.requestee` feels a little silly... `request.recipient`?

File: Gemfile [hunk 3]
```diff
@@ -36,7 +36,12 @@ gem 'jbuilder', '~> 2.5'
 # Reduces boot times through caching; required in config/boot.rb
 gem 'bootsnap', '>= 1.1.0', require: false
 
+# Authentication
 gem 'devise', '~> 4.6.2'
+gem 'devise_invitable', '~> 2.0.0'
+
+# Authorization
+gem 'pundit'
```
Comment (Br3nda) [inline Gemfile]: :tada:

File: app/controllers/organisation_member/requests_controller.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+class OrganisationMember::RequestsController < OrganisationMember::BaseController
+  before_action :set_request, only: :show
+
+  # GET /requests
+  def index
+    @requests = @organisation.requests
+  end
+
+  def new
+    @request = Request.new(requester: @organisation)
```
Comment (Br3nda) [inline app/controllers/organisation_member/requests_controller.rb]: needs a test for this method

File: spec/controllers/organisation_member/dashboard_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+RSpec.describe OrganisationMember::DashboardController do
+  let(:organisation_member) { FactoryBot.create(:organisation_member) }
+  let(:user) { organisation_member.user }
+  let(:organisation) { organisation_member.organisation }
+
+  before do
+    sign_in user
+  end
+
+  describe 'GET index' do
```
[the remark this answers is not in the document]
Comment (mermop) [reply spec/controllers/organisation_member/dashboard_controller_spec.rb]: I'm not too concerned about duplication across specs

File: Gemfile (hunk 3 above)
Comment (mermop) [reply Gemfile]: It was already in here, I just moved it up!

File: app/controllers/organisation_member/requests_controller.rb (hunk 4 above)
Comment (mermop) [reply app/controllers/organisation_member/requests_controller.rb]: done in 6d98368

File: spec/controllers/organisation_member/requests_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        get :show, params: { organisation_id: organisation.id, id: another_request.id }
+      end.to raise_error(ActiveRecord::RecordNotFound)
+    end
+  end
+  describe 'GET new' do
+    it 'creates a request with the organisation as the requester' do
+      get :new, params: { organisation_id: organisation.id }
+      expect(assigns(:request).requester).to eq(organisation)
+    end
+  end
```
Comment (Br3nda) [inline spec/controllers/organisation_member/requests_controller_spec.rb]: how about a test for #create as well?

Comment (mermop) [reply spec/controllers/organisation_member/requests_controller_spec.rb]: This is being covered by the feature spec and I'm keen to keep moving

**Pull request merged**